bot/cogs/verification.py

```diff
@@ -32,26 +59,372 @@
+            f"be kicked as they haven't verified in `{KICKED_AFTER}` days. This is `{percentage:.2%}` of the "
+            f"guild's population. Proceed?",
+            allowed_mentions=MENTION_CORE_DEVS,
+        )
+
+        options = (constants.Emojis.incident_actioned, constants.Emojis.incident_unactioned)
+        for option in options:
+            await confirmation_msg.add_reaction(option)
+
+        def check(reaction: discord.Reaction, user: discord.User) -> bool:
+            """Check whether `reaction` is a valid reaction to `confirmation_msg`."""
+            return (
+                reaction.message.id == confirmation_msg.id  # Reacted to `confirmation_msg`
+                and str(reaction.emoji) in options  # With one of `options`
+                and not user.bot  # By a human
+            )
+
+        timeout = 60 * 5  # Seconds, i.e. 5 minutes
+        try:
+            choice, _ = await self.bot.wait_for("reaction_add", check=check, timeout=timeout)
+        except asyncio.TimeoutError:
+            log.debug("Staff prompt not answered, aborting operation")
+            return False
+        finally:
+            await confirmation_msg.clear_reactions()
+
+        result = str(choice) == constants.Emojis.incident_actioned
+        log.debug(f"Received answer: {choice}, result: {result}")
+
+        # Edit the prompt message to reflect the final choice
+        await confirmation_msg.edit(
+            content=f"Request to kick `{n_members}` members was {'authorized' if result else 'denied'}!"
+        )
+        return result
+
+    async def _kick_members(self, members: t.Set[discord.Member]) -> int:
+        """
+        Kick `members` from the PyDis guild.
+
+        Note that this is a potentially destructive operation. Returns the amount of successful
+        requests. Failed requests are logged at info level.
+        """
+        log.info(f"Kicking {len(members)} members from the guild (not verified after {KICKED_AFTER} days)")
+        n_kicked, bad_statuses = 0, set()
+
+        for member in members:
+            if is_verified(member):  # Member could have verified in the meantime
+                continue
+            with suppress(discord.Forbidden):
+                await member.send(KICKED_MESSAGE)  # Send message while user is still in guild
+            try:
+                await member.kick(reason=f"User has not verified in {KICKED_AFTER} days")
+            except discord.HTTPException as http_exc:
+                bad_statuses.add(http_exc.status)
+            else:
+                n_kicked += 1
+
+        self.bot.stats.incr("verification.kicked", count=n_kicked)
+
+        if bad_statuses:
+            log.info(f"Failed to kick {len(members) - n_kicked} members due to following statuses: {bad_statuses}")
+
+        return n_kicked
+
+    async def _give_role(self, members: t.Set[discord.Member], role: discord.Role) -> int:
```

Comment:
Hmm I was playing with the idea but I felt like a) I wasn't actually saving much code, and b) it led to a little bit of indirection with what's happening which I really wanted to avoid for this feature. I will give it another try.

Reply:
I added a generic executor in 1958978 and then adjusted the two methods to use it in dc70a01. The idea is similar to what I was showing to you earlier, but that solution was flawed due to the fact the the built coros weren't able to communicate whether they aborted due to `is_verified(member)` giving `True`. This should no longer be a problem with this new approach, as I was able to move the check outside of the coro.

Reply:
Looks good now.